### DISCUSSION TITLE
2d constraint system - ideas

### DISCUSSION THREAD

**Discussion by Irev-Dev** [zalo/CascadeStudio#54] [Ideas]
I've been thinking for a while Code-CAD really could do with 2d-constraint sketching system. I understand this is not an easy thing to add. This is not a feature request just wanted to throw some ideas out there.
Take this part for an example:
![image](https://user-images.githubusercontent.com/29681384/103430709-e896f500-4c1a-11eb-87f1-88f957f4e379.png)

There's a lot going on so let's focus on the top view, an just the outer bound of the shape.
The outer bounds are defined by two arcs of different sizes with tangential lines between them.
![image](https://user-images.githubusercontent.com/29681384/103430912-bedecd80-4c1c-11eb-8515-4175e8aa6f72.png)

Without a 2d constraint system would require the user to do a bunch of maths to determine where exact points where the lines need to be to be tangential with the arcs. Though, with a 2d constraint system defining them is trivial because you just do what I said in English in some kind of API.
Just spit ballin' here, but maybe if you define points by giving them names, and by convention consecutive points are connected to form a closed shape in a clockwise fashion, then a user can add constraints as to the named points. Here's an example API in a jsonish/object type format:
```javascript
{
  points: ['A', 'B', 'C', 'D'],
  constraints: {
    tangent: ['A', 'B', 'C', 'D'],
    line: [['A', 'B'], ['C', 'D']],
    arc: [
      {
        points: ['B', 'C'],
        radius: 1,
        centerPointName: 'centerBC'
      },
      {
        points: ['D', 'A'],
        radius: 1.89,
        centerPointName: 'centerDA'
      }
    ],
    vertical: [['B', 'C'], ['D', 'A']]
    horizontal: [['centerBC', 'centerDA']]
    coordinates: [{
      point: 'centerDA',
      coords: [0, 0],
    }]
    horizontalDistance:[{
      points: ['centerBC', 'centerDA'],
      // assuming the radiuses were in variables this would become something like:
      // partLength - keyOuterRadius - radius2
      distance: 9.75 - (1.89 + 1) 
    }]
  }
}
// I think the above would be fully contained
```
I'm sure the above could be improved to be more expressive, it definitely a little hard to parse, but I think if point names were also able to be displayed in the 3d view under certain circumstances it would be fine (certainly a lot easier to read that the maths to get the tangents etc working without a constraint system).

Further, I stumbled across some of Michael F Bryan's work, he seems to be interested CAD development and he's built a 2d constraint solver
[Blog](https://adventures.michaelfbryan.com/posts/constraints-part-1-expressions/?utm_source=reddit&utm_medium=social&utm_campaign=constraint-solver-1-expression-trees), [reddit post](https://www.reddit.com/r/rust/comments/hr5gup/geometric_constraint_solvers_part_1_algebraic/), [github](https://github.com/Michael-F-Bryan/constraints)

It's written in rust, but since this project is already using wasm, maybe it could be useful? Worth noting I think we built a parser for taking mathematical expressions as well as the actually constraints, maybe that's better than the JSONish API I'm proposing.
Again just throwing ideas out there.

(I got the part drawings from: https://grabcad.com/library/freecad-practice-part-1)

**Comment by Irev-Dev** [comment]
If you look in the comments of the reddit post, I'm there asking how easy it would be to compile to wasm. I actually never got around to doing this but I had planned on doing some more material experiments with the ideas above, haven't found the time yet.

Actually, I should tag him. Hi @Michael-F-Bryan, not sure if this is of any interest to you.

**Reply by Michael-F-Bryan** [reply]
> I actually never got around to doing this but I had planned on doing some more material experiments with the ideas above, haven't found the time yet.

In my experience, Rust is considerably easier to compile to WebAssembly than any other language and it's `wasm-bindgen` crate makes it pretty easy to create high-level APIs that can be used from JavaScript.

For the user side of implementing a constraints system I'd recommend checking out SolidWorks. They have a mechanism called mates (for 3D) and relations (for your 2D sketches), which are essentially pre-programmed equations you can use to constrain certain properties of an object. For example, by clicking a line and a circle and telling their endpoints to be tangential.

Under the hood you reduce the relationships to a series of equations, then solve those relations to get a range of values that your object parameters are allowed to take.

> Just spit ballin' here, but maybe if you define points by giving them names, and by convention consecutive points are connected to form a closed shape in a clockwise fashion, then a user can add constraints as to the named points. 

You probably don't need to name the points. Everything would be done through the UI and internally you can just use object references or indices to refer to particular entities or entity properties.

**Reply by Irev-Dev** [reply]
Thanks @Michael-F-Bryan,
I think I should have given you a bit more context. This particular project is "Code-CAD" meaning there is no UI for the actual modelling process (similar to openscad), hence why the API was proposed.

But yeah exactly like the constraints processes in tools like Solidworks 👍, but ported to a "Code-CAD" format.

**Comment by Irev-Dev** [comment]
I've more or less decided my ideas here aren't supper well suited to Code-CAD, and I've taken a different direction with #56.
